upgrade_test.py

```diff
@@ -736,87 +746,79 @@ def test_rolling_upgrade(self):  # pylint: disable=too-many-locals,too-many-stat
+        if self.params.get('alternator_port'):
+            self.pre_create_alternator_tables()
+        if wait_for_finish is True:
+            InfoEvent(message=f"Waiting for {workload_name} to finish").publish()
+            self.verify_stress_thread(workload_thread_pool)
+        else:
+            self.log.info('Sleeping for 60s to let cassandra-stress start before the next steps...')
+            time.sleep(60)
+        return workload_thread_pool
+
+    def test_generic_cluster_upgrade(self):
         """
-        Upgrade half of nodes in the cluster, and start special read workload
+        Upgrade specified number of nodes in the cluster, and start special read workload
         during the stage. Checksum method is changed to xxhash from Scylla 2.2,
         we want to use this case to verify the read (cl=ALL) workload works
         well, upgrade all nodes to new version in the end.
-        """
-        # prepare workload (stress_before_upgrade)
-        InfoEvent(message="Starting stress_before_upgrade - aka prepare step").publish()
-        stress_before_upgrade = self.params.get('stress_before_upgrade')
-        prepare_thread_pool = self.run_stress_thread(stress_cmd=stress_before_upgrade)
-        if self.params.get('alternator_port'):
-            self.pre_create_alternator_tables()
-        InfoEvent(message="Waiting for stress_before_upgrade to finish").publish()
-        self.verify_stress_thread(prepare_thread_pool)
-
-        # Starting workload during entire upgrade
-        InfoEvent(message="Starting stress_during_entire_upgrade workload").publish()
-        stress_during_entire_upgrade = self.params.get('stress_during_entire_upgrade')
-        stress_thread_pool = self.run_stress_thread(stress_cmd=stress_during_entire_upgrade)
-        self.log.info('Sleeping for 60s to let cassandra-stress start before the rollback...')
-        time.sleep(60)
 
-        num_nodes_to_rollback = self.params.get('num_nodes_to_rollback')
-        upgraded_nodes = []
+        For multi-dc upgrades, alternates upgraded nodes between dc's.
+        """
+        step = itertools.count(start=1, step=1)
+        self._run_stress_workload("stress_before_upgrade", wait_for_finish=True)
+        stress_thread_pool = self._run_stress_workload("stress_during_entire_upgrade", wait_for_finish=False)
 
         # generate random order to upgrade
-        nodes_num = len(self.db_cluster.nodes)
-        # prepare an array containing the indexes
-        indexes = list(range(nodes_num))
-        # shuffle it so we will upgrade the nodes in a random order
-        random.shuffle(indexes)
+        nodes_to_upgrade = self.shuffle_nodes_and_alternate_dcs(list(self.db_cluster.nodes))
 
-        upgrade_sstables = self.params.get('upgrade_sstables')
+        # Upgrade all nodes that should be rollback later
+        upgraded_nodes = []
+        for node_to_upgrade in nodes_to_upgrade[0:self.params.get('num_nodes_to_rollback')]:
```

Comment:
Start index is redundant here. Can be just `[:self.params.get('num_nodes_to_rollback')]`